core/src/main/scala/sttp/tapir/Validator.scala

```diff
@@ -160,75 +200,130 @@ object Validator extends ValidatorMacros {
-      case Enumeration(possibleValues, _, _) => Some(s"in(${possibleValues.mkString(",")}")
-      case Mapped(wrapped, _)                => show(wrapped)
-      case All(validators) =>
-        validators.flatMap(show(_)) match {
-          case immutable.Seq()  => None
-          case immutable.Seq(s) => Some(s)
-          case ss               => Some(s"all(${ss.mkString(",")})")
-        }
-      case Any(validators) =>
-        validators.flatMap(show(_)) match {
-          case immutable.Seq()  => Some("reject")
-          case immutable.Seq(s) => Some(s)
-          case ss               => Some(s"any(${ss.mkString(",")})")
-        }
+    override def show: Option[String] = validators.flatMap(_.show) match {
+      case Nil      => Some("reject")
+      case x :: Nil => Some(x)
+      case xs       => Some(s"any(${xs.mkString(",")})")
     }
+
+    override def apply(t: T): ValidationResult[T] =
+      ValidationResult.partitionMap(
+        validators.toList.map(_.apply(t))
+      ) match {
+        case (_, _ :: _)    => ValidationResult.Valid(t)
+        case (invalid, Nil) => ValidationResult.Invalid(t, invalid.flatMap(_.errors))
+      }
+
+    override def contramap[TT](g: TT => T): Validator[TT] =
+      if (validators.isEmpty) Any(Nil) else super.contramap(g)
+
+    override def or(other: Validator[T]): Validator[T] =
+      if (validators.isEmpty) other else Any(validators :+ other)
   }
 }
 
-sealed trait ValidationError[T] {
-  def prependPath(f: FieldName): ValidationError[T]
-  def invalidValue: T
-  def path: List[FieldName]
+sealed trait ValidationResult[+T] {
+  def valid[TT >: T]: Option[Valid[TT]] =
+    this match {
+      case v: Valid[TT]   => Some(v)
+      case _: Invalid[TT] => None
+    }
+
+  def invalid[TT >: T]: Option[Invalid[TT]] =
+    this match {
+      case _: Valid[TT]   => None
+      case i: Invalid[TT] => Some(i)
+    }
 }
+object ValidationResult {
+  case class Valid[T](value: T) extends ValidationResult[T]
+  case class Invalid[T](value: T, errors: List[ValidationError]) extends ValidationResult[T] {
 
-object ValidationError {
-  case class Primitive[T](validator: Validator.Primitive[T], invalidValue: T, path: List[FieldName] = Nil) extends ValidationError[T] {
-    override def prependPath(f: FieldName): ValidationError[T] = copy(path = f :: path)
+    /** Messages describing a list of validation errors: which values are invalid, and why. */
+    def allMessages: List[ValidationMessage] =
+      errors.map(_.message)
+
+    /** Message describing a list of validation errors: which values are invalid, and why. */
+    def description: String =
+      allMessages.map(_.value).mkString(",")
   }
 
-  case class Custom[T](invalidValue: T, message: String, path: List[FieldName] = Nil) extends ValidationError[T] {
-    override def prependPath(f: FieldName): ValidationError[T] = copy(path = f :: path)
+  def partitionMap[A](xs: List[ValidationResult[A]]): (List[Invalid[A]], List[Valid[A]]) = {
```

Comment:
shouldn't this be called simply `parition`, as there's no mapping here?

Reply:
For sure!